### ISSUE TITLE
`reserveFunds` flickers in integration test

### ISSUE THREAD

**Issue by andrewgordstewart** [statechannels/monorepo#1478]
[This job](https://app.circleci.com/pipelines/github/statechannels/monorepo/5227/workflows/33472b40-ff5a-469e-80c7-e4e9c0851bda/jobs/19408) failed on https://github.com/statechannels/monorepo/pull/1475/commits/9933371e563bf36fde63c0dc4e1e2ff90b441a74 of https://github.com/statechannels/monorepo/pull/1475.

The change did nothing to directly affect `xstate-wallet`. Both integration tests passed on the previous commit https://github.com/statechannels/monorepo/pull/1475/commits/559e5a0cc1cf4d5ef84e5d517a418a76d6c66d9f.

The [browser logs](https://19408-209829093-gh.circle-artifacts.com/0/tmp/seed-download.browser.log) reveal an error occurred in the `reserveFunds` service.

**Comment by andrewgordstewart** [comment]
Going to use [this test failure](https://app.circleci.com/pipelines/github/statechannels/monorepo/5260/workflows/fbf9fb00-6b13-4d0d-b026-018921d48182/jobs/19597).

**Comment by andrewgordstewart** [comment]
Note: [this failure](https://app.circleci.com/pipelines/github/statechannels/monorepo/5260/workflows/fbf9fb00-6b13-4d0d-b026-018921d48182/jobs/19597) has the unique request id fix in its history.

**Comment by andrewgordstewart** [comment]
Here's what I found:
```
# https://app.circleci.com/pipelines/github/statechannels/monorepo/5260/workflows/fbf9fb00-6b13-4d0d-b026-018921d48182/jobs/19597/artifacts
https://github.com/statechannels/monorepo/issues/1478

cat seed-download.pino.log | jq -s -c 'sort_by(.time) | .[]' > sorted.log

# Sort lines by line length
cat sorted.log | yarn pino-pretty > pretty.log

# The error was thrown by browser 1:
# https://19597-209829093-gh.circle-artifacts.com/0/tmp/seed-download.browser.log
cat sorted.log | grep '"browserId":1' > browser1.log
cat browser1.log | pino-pretty > browser1.pretty.log

# This is the application channel missing the application site
cat browser1.log | grep 0xb07f6ec88879f4d928861c57be66bb47d492a0b04549c51402dbadfb36f560e6 | pino-pretty > channels.pretty.log

# The error happened _after_ the site was "successfully" set
cat browser1.log | grep 'settingSite' >  settingSite.log

# The other log activity around that same time, that could have touched the store, was the resolution
# of the pushMessage promise:
cat browser1.log | grep '"id":7' > pushMessage.log  

# Combining these two:
❯ cat problem.log| pino-pretty -t 

[2020-04-20 13:37:57.273 +0000] INFO  (xstate-wallet): INCOMING JSONRPC REQUEST:
    jsonRpcRequest: {
      "jsonrpc": "2.0",
      "method": "PushMessage",
      "params": {
        "recipient": "0xb562118afF1b666601Ca6524018729d06B30c489",
        "sender": "0xEa61e52e777cCD897d0ddD755eb28d12a8529356",
        "data": {
          "signedStates": [
            {
              "challengeDuration": "0x000000000000000000000000000000000000000000000000000000000000012c",
              "channelNonce": "0x0000000000000000000000000000000000000000000000000000000000000000",
              "turnNum": "0x0000000000000000000000000000000000000000000000000000000000000000",
              "outcome": [
                {
                  "assetHolderAddress": "0x4ad3F07BEFDC54511449A1f553E36A653c82eA57",
                  "allocationItems": [
                    {
                      "destination": "0x000000000000000000000000D9995BAE12FEe327256FFec1e3184d492bD94C31",
                      "amount": "0x0000000000000000000000000000000000000000000000000000000000000000"
                    },
                    {
                      "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39",
                      "amount": "0x0000000000000000000000000000000000000000000000000000000000b71b00"
                    }
                  ]
                }
              ],
              "channelId": "0xb07f6ec88879f4d928861c57be66bb47d492a0b04549c51402dbadfb36f560e6",
              "signatures": [
                "0x528cf98656bacd5a42ea2ce774de734f29b091298782d6fe6f7b0f3f1bf6b8e0427b76a320e4ceed77196e048b2dcba95d7ec331bc7f4d56b23dbfa68f09c2b61b"
              ],
              "appData": "0x0000000000000000000000000000000000000000000000000000000000000000",
              "appDefinition": "0x430869383d611bBB1ce7Ca207024E7901bC26b40",
              "isFinal": false,
              "chainId": "9001",
              "participants": [
                {
                  "participantId": "0xEa61e52e777cCD897d0ddD755eb28d12a8529356",
                  "signingAddress": "0xEa61e52e777cCD897d0ddD755eb28d12a8529356",
                  "destination": "0x000000000000000000000000D9995BAE12FEe327256FFec1e3184d492bD94C31"
                },
                {
                  "participantId": "0xb562118afF1b666601Ca6524018729d06B30c489",
                  "signingAddress": "0xb562118afF1b666601Ca6524018729d06B30c489",
                  "destination": "0x000000000000000000000000d4Fa489Eacc52BA59438993f37Be9fcC20090E39"
                }
              ]
            }
          ]
        }
      },
      "id": 7
    }
    browserId: 1
[2020-04-20 13:37:57.299 +0000] INFO  (xstate-wallet): JOIN_CHANNEL-0xb07f6ec88879f4d928861c57be66bb47d492a0b04549c51402dbadfb36f560e6 - TRANSITION FROM {"joiningChannel":"joining"} EVENT JOIN_CHANNEL TO {"joiningChannel":"settingSite"}
    browserId: 1
[2020-04-20 13:37:57.318 +0000] INFO  (xstate-wallet): JOIN_CHANNEL-0xb07f6ec88879f4d928861c57be66bb47d492a0b04549c51402dbadfb36f560e6 - TRANSITION FROM {"joiningChannel":"settingSite"} EVENT done.invoke.setApplicationSite TO 'fundingChannel'
    browserId: 1
[2020-04-20 13:37:57.325 +0000] INFO  (xstate-wallet): OUTGOING JSONRPC REQUEST:
    jsonRpcResponse: {
      "id": 7,
      "jsonrpc": "2.0",
      "result": {
        "success": true
      }
    }
    browserId: 1

```

Note that the outgoing RPC request was logged _after_ `done.invoke.setApplicationSite`. So, [this promise](https://github.com/statechannels/monorepo/blob/master/packages/xstate-wallet/src/workflows/application.ts#L349) resolved between  time `2020-04-20 13:37:57.299 +0000` and `2020-04-20 13:37:57.318 +0000`.

Store functions which modify channels are supposed to [acquire a lock](https://github.com/statechannels/monorepo/blob/master/packages/xstate-wallet/src/store/store.ts#L295-L297) before fetching the channel. Otherwise, they run the risk of having their update clobbered by a concurrent write.

[`setApplicationSite`](https://github.com/statechannels/monorepo/blob/master/packages/xstate-wallet/src/store/store.ts#L287-L293) does not acquire a lock. So, I feel fairly confident that this is what happened.
